### ISSUE TITLE
Hints about foldr vs. foldl (and related) when combined with reverse

### ISSUE THREAD

**Issue by jvoigtlaender** [ndmitchell/hlint#1057]
I'd like to add the following hints to either the `default` group or the `teaching` group. I'm not sure in which of these two groups they should go, hence opening this as an issue rather than pull request for now:
```yaml
- warn: {lhs: foldr f c (reverse x), rhs: foldl (flip f) c x, name: Use foldl instead of foldr}
- warn: {lhs: foldr1 f (reverse x), rhs: foldl1 (flip f) x, name: Use foldl1 instead of foldr1}
- warn: {lhs: foldl f c (reverse x), rhs: foldr (flip f) c x, note: IncreasesLaziness, name: Use foldr instead of foldl}
- warn: {lhs: foldl1 f (reverse x), rhs: foldr1 (flip f) x, note: IncreasesLaziness, name: Use foldr1 instead of foldl1}
```

**Comment by ndmitchell** [comment]
As a first approximation, no one should be using foldl, but instead using the `foldl'` variant which eliminates the space leak. Does replacing foldl of reverse potentially introduce a space leak? Replacing foldr/reverse with foldl is definitely a good thing to do always. Do we have a rule that suggests foldl to foldl'? Should we?

I think these hints belong in the normal collection - they're useful.

**Comment by jvoigtlaender** [comment]
First off, I hadn't thought about `foldl'` at all, but yes, I see that it is relevant.

As far as I can see, there is no existing rule suggesting `foldl -> foldl`.

There is at least one existing rule that introduces `foldl`: https://github.com/ndmitchell/hlint/blob/fe9c3ad66968840fcd61a9b967ffd9a89ff2970d/data/hlint.yaml#L226

I wonder whether that one, and some of my new proposed rules, should indeed suggest `fold'`. That could be problematic, since for some choices of `f` the use of `foldl'` would be detrimental, in the sense of turning a terminating program into a nonterminating one.

So I guess one question to be answered up front is whether consideration of `foldl` vs. `foldl'` belongs in individual rules that (disregarding space leaks) happen to involve `foldl` or whether it should be handled separately.

**Comment by ndmitchell** [comment]
It seems like `foldl` to `foldl'` should be a separate rule. Therefore, I have two concerns with these rules:

* If we suggest foldl/reverse => foldr are we making it harder for the user to then get the right strictness? A more judgement question, but if technically we potentially eliminate the space leak, that would be good enough to move forward.
* If we suggest foldr/reverse => foldl does that introduce a space leak or preserve the status quo? A technical question, but I'm not sure enough to know what the answer is.

**Comment by jvoigtlaender** [comment]
A possible third question:
- Should there also be a rule `foldl' f c (reverse x) -> foldr (flip f) c x`? (In case it is worked out that "`reverse` already has a space leak, and using `foldr` doesn't make that space leak asymptotically worse")

**Comment by ndmitchell** [comment]
Yep. Sounds like we have three space leaks to figure out, and then we can move forward with this issue.

**Comment by jvoigtlaender** [comment]
Okay, let me try my hand at this "figuring out" for the first of the three cases, also to see/get feedback what kind of reasoning is even relevant in these considerations:

Say we have `foldl f c (reverse (replicate n undefined))`; the `replicate`-call because it represents a list of length `n` but does only take up constant space at the beginning, producing list elements only on demand.

Starting from `foldl f c (reverse (replicate n undefined))`, we first have to fully evaluate the `reverse`-call before `foldl` can make any progress. That `reverse` call will enforce that indeed the `replicate`-call is unfolded `n` or `n+1` times, and creates a list of length `n` that needs to be fully stored in memory at that point (it cannot be partially consumed). Only afterwards, `foldl` can start its work on that list as a whole. It will create the nested applications `f c undefined`, `f (f c undefined) undefined` etc., and cannot evaluate any of them before reaching the end of the list. So this again creates a whole tree without partial reductions along the way. Both the intermediate list and that tree have space requirement O(n). Ultimately, evaluation of the tree `f (f (f ...) ...) ...` kicks in, and what happens then depends on strictness properties of `f` (does it evaluate its first argument, or throw it away, ...?).

Now, what if we had applied the candidate rule and replaced the original definition by `foldr (flip f) c (replicate n undefined)`? Now the `replicate` can actually be performed interleaved with the `foldr`. That is, `replicate` will produce elements on demand that are immediately consumed by `foldr`. The intermediate list does not need to be represented as a whole in memory at any point. Nevertheless, space usage might be O(n) - but not worse - depending on what properties `f` has. Since now evaluation proceeds roughly as follows:
```
   foldr (flip f) c (replicate n undefined)
=> f (foldr (flip f) c (replicate (n - 1) undefined)) undefined
{- either reducing the outer f now, if it only needs its second argument, or proceeding as follows -}
=> f (f (foldr (flip f) c (replicate (n - 2) undefined)) undefined) undefined
=> ...
```
So if `f` ignores its first argument, we can short-circuit, but if it doesn't, we have to go down the left spine while storing all the right children, which ends up consuming O(n) space - possibly with a lower constant factor than before, though.

In that sense, I think applying the rule would be justified.

Relevant now:
- Did the above even address the relevant question?
- Is the reasoning correct?
- Are there other aspects to consider?

And if the above is a sufficient argument, how do the analogous considerations play out for the other two candidate rules? My gut feeling is that all three rules are worthwhile to apply, because in all of them the `reverse` call present originally already has any space leak there could be, because it has to fully produce an in-memory representation of a list of length `n`. Any space needs of the other functions can only be on the same O(n) order, and sometimes they might even be cut out, or the space leak from the `reverse` actually being eliminated.

**Comment by ndmitchell** [comment]
I think I buy the argument that it's no more or less of a space leak. But from `foldr` it's way harder to get to `foldl'`. What I don't know is if `foldl'` with `reverse` is less of a space leak than the foldr` without the `reverse`?

**Comment by jvoigtlaender** [comment]
That is more or less equivalent to the question whether `foldl' f c (reverse x) -> foldr (flip f) c x` would be a worthwhile rule to apply, right? (Even more so if hlint had a rule proposing `foldl'` for `foldl`.)

I think asymptotically the space leaks on the left and on the right of this are the same, but in terms of constant factors the right-hand side could amass more allocations, because what the `foldr` allocates for each recursive step could be a larger constant than what the `reverse` allocates for each recursive step, namely "just" a cons-cell (`:`).

So, maybe having any of these rules in the `default` section is premature.

What about adding them to the `teaching` section? In my specific case, at the point in teaching where I come upon these functions, I only care about expressiveness, and recursion patterns, and understanding how higher-order functions work. I don't even tell the students about `foldl'`, our naive cost model (as far as one exists at all) at that point is "counting number of lazy evaluation steps". That would justify having the rules `foldr/reverse -> foldl` and `foldl/reverse -> foldr`. But that is just my course. I could also add these two rules to our local setup, without any change to hlint at the moment. (Maybe revisiting once `foldl'` is handled/proposed by hlint in appropriate cases.)

**Comment by josephcsible** [comment]
> As a first approximation, no one should be using foldl, but instead using the `foldl'` variant which eliminates the space leak. Does replacing foldl of reverse potentially introduce a space leak? Replacing foldr/reverse with foldl is definitely a good thing to do always. Do we have a rule that suggests foldl to foldl'? Should we?

`foldl'` isn't always better than `foldl`. Consider `data SnocList a = Nil | Snoc (SnocList a) a` or some other left-associative type. For them, `foldl` is good and `foldr` has a space leak and should be replaced with `foldr'`. As such, I don't think we should have a hint between the strict and lazy versions unless we somehow know the type they're being used at.

**Comment by ndmitchell** [comment]
The number of legitimate `foldl` uses is so low that I suggest all of them should have a comment so some well-meaning human doesn't go in an inadvertently change them to `foldl'`. Once you are at that point, having them a default hint in HLint seems fine.

For these hints, teaching seems absolutely fine - they are great teaching hints, as people should learn the duality. Given all the concerns around space leaks, and general unease, my inclination is to only put them in teaching for now, and consider moving them later.

**Comment by jvoigtlaender** [comment]
@josephcsible, since you brought up `foldr'`, of whose existence I was previously ignorant, I'd be interested in your take on the rules in https://github.com/ndmitchell/hlint/pull/1099.

Moreover, taking into account the concern of suggesting `foldl'` (or indeed `foldr'`) on unknown types, how about in the case of a known type? For rules of the `fold/reverse` kind, we do know the type of operation, since unlike the folds, `reverse` is only defined on lists. So would you be in favor of rules like the following ones?
```yaml
    - warn: {lhs: foldr f c (reverse x), rhs: foldl' (flip f) c x, name: Replace fold by strict fold}
    - warn: {lhs: foldl f c (reverse x), rhs: foldr' (flip f) c x, name: Replace fold by strict fold}
```
I am a bit skeptical about them myself, since for certain `f`-definitions they might make a program less defined on some inputs.

Apart from those thoughts, I guess this issue here can be closed.

**Comment by josephcsible** [comment]
I'm in favor of those kinds of rules. Two minor quibbles:

1. You say "for certain `f`-definitions they might make a program less defined on some inputs", but I'm not convinced that's true. Do you have an example?
2. That's not an efficient use of `foldr'`. I'd use these rules instead:
```yaml
    - warn: {lhs: foldr f c (reverse x), rhs: foldl' (flip f) c x}
    - warn: {lhs: foldr' f c (reverse x), rhs: foldl' (flip f) c x}
    - warn: {lhs: foldl f c (reverse x), rhs: foldr (flip f) c x}
    - warn: {lhs: foldl' f c (reverse x), rhs: foldr (flip f) c x}
```

**Comment by jvoigtlaender** [comment]
About 1.:
```
Prelude Data.List> let f = curry fst
Prelude Data.List> let c = undefined
Prelude Data.List> let x = [undefined,True]
Prelude Data.List> foldr f c (reverse x)
True
Prelude Data.List> foldl' (flip f) c x
*** Exception: Prelude.undefined
CallStack (from HasCallStack):
  error, called at libraries\base\GHC\Err.hs:78:14 in base:GHC.Err
  undefined, called at <interactive>:24:9 in interactive:Ghci11
```
I'm pretty sure something similar can be done to the other case.

**Comment by jvoigtlaender** [comment]
About 2., the four rules there:

These ones:
```yaml
    - warn: {lhs: foldr' f c (reverse x), rhs: foldl' (flip f) c x}
    - warn: {lhs: foldl f c (reverse x), rhs: foldr (flip f) c x}
```
were added in https://github.com/ndmitchell/hlint/commit/55da832da1bbcb15d3517391bef2bc9e7307f7a9 and https://github.com/ndmitchell/hlint/commit/65d9279cd24031c4dbf8616a4ff7aee288fc44b3.

This one:
```yaml
    - warn: {lhs: foldr f c (reverse x), rhs: foldl' (flip f) c x}
```
is one of those that can lead to less-definedness.

This one:
```yaml
    - warn: {lhs: foldl' f c (reverse x), rhs: foldr (flip f) c x}
```
I already wondered about further above (https://github.com/ndmitchell/hlint/issues/1057#issuecomment-648747655), but that discussion did not lead to a conclusion.

In general, would you say that `foldr'` is never efficient/worthwhile to use on lists?

**Comment by josephcsible** [comment]
Ah, that's a clever example. So since `foldl'` is strict in the accumulator, it can't be undefined at any intermediate step even if it's never used. A "real-world" consequence of this is that `foldr const undefined . reverse` is a valid implementation of `last`, as is `foldl (flip const) undefined`, but `foldl' (flip const) undefined` is not.

**Comment by jvoigtlaender** [comment]
Closing this, since I think only https://github.com/ndmitchell/hlint/pull/1102 is the only action item left from all the above at the moment (before more experience is accumulated and maybe leads to consideration of such rules outside the teaching group).
